File: backend/src/main/kotlin/taskmanager/backend/models/Comment.kt
package taskmanager.backend.models

import org.bson.types.ObjectId
import taskmanager.backend.dtos.response.CommentResponseDto
import taskmanager.backend.models.base.impl.BaseEntityImpl
import taskmanager.backend.models.interfaces.AttachedToTaskEntity
import taskmanager.backend.models.interfaces.MappableCreatedByUserEntity

data class Comment(
    override val createdBy: ObjectId,
    override val task: ObjectId,

    var text: String
) : BaseEntityImpl(), AttachedToTaskEntity, MappableCreatedByUserEntity<CommentResponseDto> {

    override fun toResponseDto(userId: ObjectId, createdBy: User?): CommentResponseDto {
        return CommentResponseDto(
            _id = _id.toString(),
            createdBy = createdBy?.toResponseDto(),
            task = task.toString(),
            text = text,
            createdAt = createdAt,
            updatedAt = updatedAt,
            canEdit = userId.toString() == userId.toString()
        )
    }
}

File: frontend/lib/models/comment.dart
import 'package:frontend/models/base/base_entity.dart';
import 'package:frontend/models/user.dart';

class Comment extends BaseEntity {
  User? createdBy;
  String task;
  String text;
  bool canEdit;

  Comment({
    super.id,
    this.createdBy,
    this.task = '',
    this.text = '',
    super.createdAt,
    super.updatedAt,
    this.canEdit = false,
  });

  factory Comment.fromJson(Map<String, dynamic> json) {
    return Comment(
      id: json['_id'],
      createdBy: json['createdBy'] != null ? User.fromJson(json['createdBy'] as Map<String, dynamic>) : null,
      task: json['task'],
      text: json['text'],
      createdAt: BaseEntity.parseDateFromJson(json['createdAt']),
      updatedAt: BaseEntity.parseDateFromJson(json['updatedAt']),
      canEdit: json['canEdit'],
    );
  }
}
